docs: Component Rules (#1055)
deephaven/deephaven-plugins: plugins/ui/docs/describing/component_rules.md

# Component Rules

Here are some important rules to understand about `deephaven.ui` components.

## Children and Props

> Arguments passed to a component may be either `children` or `props`. `Children` refers to `child` components that are passed to a `parent` component as arguments. `Props` are properties that determine the behavior and rendering style of the component.

Author: Added definitions and a JSX example.

# Component Rules

> Here are some important rules to understand about `deephaven.ui` components.

Reviewer: ```suggestion
This guide presents some important rules to remember for `deephaven.ui` components when writing queries.
```

## Children and props

> Arguments passed to a component may be either `children` or `props`. `Children` refers to `child` components that are passed to a `parent` component as positional arguments. `Props` are properties passed as keyword arguments that determine the behavior and rendering style of the component. Positional arguments must be included in the correct order. Keyword arguments are included with a keyword and equals sign.

Reviewer: Worth noting that keyword argument ordering doesn't matter, but that ordering of keyword arguments can affect ordering of positional arguments and make them out of order?

Author: Added

> ```python

Reviewer: Image of the output?

Reviewer: @dsmmcken will this be testable with the snapshot tool if this doesn't live in docs?

Author: Added

from deephaven import ui

my_flex = ui.flex(
    ui.heading("Heading"),
    ui.button("Button"),
    ui.text("Text"),
    direction="column",
    wrap=True,
    width="200px",
)
```

In the above example, the `flex` component is the `parent`. It has three `children`: a `heading`, a `button`, and a `text` component. These `children` will be rendered inside the `flex`. It also has three props: `direction`, `wrap`, and `width`. These three props indicate that the flex should be rendered as a 200 pixel column with wrap enabled.

> ## Comparison with JSX

Reviewer: I don't understand the purpose of this section. What component rule does this demonstrate?

Author: @mofojed Requested a comparison with JSX.

Reviewer 2: The purpose is showing users familiar with React how to modify this for DH, but an intro sentence saying as much would be useful.

Here is a component written in React JSX.

```html
<MyComponent prop1="value1">Hello World</MyComponent>
```

Here is the same component written in `deephaven.ui`.

```python
my_component("Hello World", prop1="value1")
```

> ## Defining Your Own Children and Props

Reviewer 2: sentence case for headings, please

Author: fixed

> ## Defining your wwn children and props

Reviewer 2: ```suggestion
## Defining your own children and props
```

To define `children` and `props` for a custom component, add them as arguments to the component function. As a convention, you may declare the children using the `*` symbol to take any number of arguments.

> ```python

Reviewer: An image of the component would be good to have.

Author: Added

from deephaven import ui

@ui.component
def custom_flex(*children, is_column):
    return ui.flex(
        ui.heading("My Component"),
        children,
        direction="column" if is_column else "row",
    )

my_custom_flex = custom_flex(ui.text("text"), ui.button("button"), is_column=True)
```

## Component return values

> A `deephaven.ui` component usually returns a component. It may also return a list or tuple of components. It may return `None` if it should perform logic but does not need to be rendered. It may also return a single value like a `string` or `int`.

Reviewer: Does this differ at all from any other Python UDF?

Author: Yes. A Python UDF can return any value. A deephaven.ui component is limited in valid return values.

```python
from deephaven import ui

@ui.component
def return_component():
    return ui.text("component")

@ui.component
def list_of_components():
    return [ui.text("list"), ui.text("of"), ui.text("components")]

@ui.component
def return_tuple():
    return (ui.text("a"), ui.text("b"))

@ui.component
def return_none():
    print("return none")
    return None

@ui.component
def return_string():
    return "string"

@ui.component
def return_int():
    return 1

my_return_component = return_component()
my_list_of_components = list_of_components()
my_return_tuple = return_tuple()
my_return_none = return_none()
my_return_string = return_string()
my_return_int = return_int()
```

> ## Conditional return

Reviewer: This section feels like the same topic as the previous. Any reason not to just combine them?

Author: The previous section is about valid return values. This section is about returning different values based on a condition. I think it is clearer in two different sections.

Return statements can be conditional in order to render different components based on inputs.

```python
from deephaven import ui

@ui.component
def return_conditional(is_button):
    if is_button:
        return ui.button("button")
    return ui.text("text")

my_button = return_conditional(True)
my_text = return_conditional(False)

> ```

Reviewer: Related documentation?

Author: I can link to conditional rendering. All that section has not been created yet.

Reviewer 2: The plugins / deephaven.ui docs don't have a Related docs section; so either all of them need it, or we agree not to include it for now.

Reviewer: All of the docs in this repo will eventually be rendered on the docs site. All of the pages on there have related documentation at the bottom. These should have it.

# Component Rules

This guide presents some important rules to remember for `deephaven.ui` components when writing queries.

## Children and props

> Arguments passed to a component may be either `children` or `props`. `Children` refers to `child` components that are passed to a `parent` component as positional arguments. `Props` are properties passed as keyword arguments that determine the behavior and rendering style of the component. Positional arguments must be included in the correct order. Keyword arguments are included with a keyword and equals sign. Keyword argument ordering does not matter, but the ordering of keyword arguments can affect ordering of positional arguments and make them out of order.

Reviewer 2: ```suggestion
Arguments passed to a component may be either `children` or `props`. `Children` refers to `child` components that are passed to a `parent` component as positional arguments. `Props` are properties passed as keyword arguments that determine the behavior and rendering style of the component. Positional arguments must be included in the correct order. Keyword arguments are included with a keyword and equals sign. Keyword argument ordering does not matter, but the ordering of keyword arguments can affect the ordering of positional arguments and make them out of order.
```

This is a lot of information packed into one paragraph. Are children and parents positional arguments or keyword arguments?

## Comparison with JSX

Here is a component written in React JSX.

```html
<MyComponent prop1="value1">Hello World</MyComponent>
```

Here is the same component written in `deephaven.ui`.

```python
my_component("Hello World", prop1="value1")
```

## Define your own children and props

To define `children` and `props` for a custom component, add them as arguments to the component function. As a convention, you may declare the children using the `*` symbol to take any number of arguments.

```python
from deephaven import ui

@ui.component
def custom_flex(*children, is_column):
    return ui.flex(
        ui.heading("My Component"),
        children,
        direction="column" if is_column else "row",
    )

my_custom_flex = custom_flex(ui.text("text"), ui.button("button"), is_column=True)
```

![Define your own children and props](../_assets/component_rules_2.png)

## Component return values

> A `deephaven.ui` component usually returns a component. It may also return a list or tuple of components. It may return `None` if it should perform logic but does not need to be rendered. It may also return a single value like a `string` or `int`.

Reviewer 2: ```suggestion
A `deephaven.ui` component usually returns a component. However, it may also return:
- a list or tuple of components. 
- `None` if it should perform logic but does not need to be rendered. 
- a single value like a `string` or `int`.
```

```python
from deephaven import ui

@ui.component
def return_component():
    return ui.text("component")

@ui.component
def list_of_components():
    return [ui.text("list"), ui.text("of"), ui.text("components")]

@ui.component
def return_tuple():
    return (ui.text("a"), ui.text("b"))

@ui.component
def return_none():
    print("return none")
    return None

@ui.component
def return_string():
    return "string"

@ui.component
def return_int():
    return 1

my_return_component = return_component()
my_list_of_components = list_of_components()
my_return_tuple = return_tuple()
my_return_none = return_none()
my_return_string = return_string()
my_return_int = return_int()
```

## Conditional return

Return statements can be conditional in order to render different components based on inputs.

```python
from deephaven import ui

@ui.component
def return_conditional(is_button):
    if is_button:
        return ui.button("button")
    return ui.text("text")

my_button = return_conditional(True)
my_text = return_conditional(False)
```

> See [Conditional Rendering](./conditional_rendering.md)

Reviewer 2: ```suggestion
```

```python
from deephaven import ui

my_flex = ui.flex(
    ui.heading("Heading"),
    ui.button("Button"),
    ui.text("Text"),
    direction="column",
    wrap=True,
    width="200px",
)
```

In the above example, the `flex` component is the `parent`. It has three `children`: a `heading`, a `button`, and a `text` component. These `children` will be rendered inside the `flex`. It also has three props: `direction`, `wrap`, and `width`. These three props indicate that the flex should be rendered as a 200 pixel column with wrap enabled.

## Defining Your Own Children and Props

To define `children` and `props` for a custom component, add them as arguments to the component function. As a convention, you may declare the children using the `*` symbol to take any number of arguments.

```python
from deephaven import ui

@ui.component
def custom_flex(*children, is_column):
    return ui.flex(
        ui.heading("My Component"),
        children,
        direction="column" if is_column else "row",
    )

my_custom_flex = custom_flex(ui.text("text"), ui.button("button"), is_column=True)
```

## Component Return Values

> You can return three values from a `deephaven.ui` component: a component, a list of components, or `None`. Returning a single component will render that component at the root level of a panel. Returning a list of components will render all the components in a panel. Returning `None` should be used when a component is used to perform logic but does not need to be rendered.

Reviewer 3: @mofojed this isn't true? For instance you can also return hooks.

Author: I did not know you could return hooks from a component. Can you also return primitive values like strings, numbers, and booleans?

Reviewer 4: @dsmmcken What do you mean return hooks? You can return the result of a hook, but returning the hook itself doesn't make sense (ditto in React). The result of the hook could be a component, list of components, `None`, or a primitive.
You can also return primitives though, such as `str`, `int`, `boolean`.

Author: I pushed an update. I'm still not sure if this is saying what we want.

Author: Also, I tested dict and set. They do not open up a panel.
With bool, it opens a panel, but does not render anything (where string and int get rendered).

Reviewer 4: Yes that's correct. I don't know _why_ you'd want to return a `boolean` (since it doesn't get rendered), but React does allow you to do that. 
Returning a dict or a set could possibly give you an error. There's a little playground at the bottom of the React page where you can play with what are valid returns, we should match that: https://react.dev/learn/your-first-component

Author: Is there a way to add a code playground in our docs? Can we do that in an `md` file?

Reviewer 4: We want to do that, it will not be in the first deployment.
With React Playground it's nice because it all runs client side. We need to have a server running for our playground, so setup is a little more complicated. Probably use a running instance that's shared between all viewers of the docs.

Reviewer 3: > What do you mean return hooks? You can return the result of a hook,

Yeah, that's what I meant, like return [my_state, my_set_state]

Reviewer 4: @dsmmcken doing `return [my_state, my_set_state]` does not make sense from a component. You can do that from other hooks, but that doesn't make sense in React or in deephaven.ui.
You can do `return my_state`. You wouldn't return the setter function.
